File: cpp-thin-client/src/moderation.cpp
#include "ironwall/moderation.hpp"
#include <iostream>

namespace ironwall {

ModerationEngine::ModerationEngine(RateLimitConfig cfg) : cfg_(std::move(cfg)) {}

bool ModerationEngine::allow(const std::string& player_id, const std::string& event_type) {
    std::lock_guard lock(mtx_);
    auto now = Clock::now();

    // expire ban?
    if (auto it = bans_.find(player_id); it != bans_.end()) {
        if (now < it->second) {
            std::cerr << "[mod] DENY banned player=" << player_id
                      << " reason=" << ban_reasons_[player_id] << "\n";
            return false;
        }
        bans_.erase(it);
        ban_reasons_.erase(player_id);
    }

    auto& q = windows_[player_id];
    auto window = std::chrono::duration<double>(cfg_.window_sec);
    while (!q.empty() && (now - q.front()) > window) q.pop_front();
    q.push_back(now);

    if (q.size() > cfg_.max_events) {
        size_t& s = strikes_[player_id];
        ++s;
        std::cerr << "[mod] RATE LIMIT player=" << player_id
                  << " event=" << event_type
                  << " count=" << q.size()
                  << " strike=" << s << "\n";
        if (s >= cfg_.ban_after_violations) {
            bans_[player_id] = now + std::chrono::duration_cast<Clock::duration>(
                std::chrono::duration<double>(cfg_.ban_duration_sec));
            ban_reasons_[player_id] = "rate_limit";
            std::cerr << "[mod] BAN player=" << player_id << " for "
                      << cfg_.ban_duration_sec << "s\n";
        }
        return false;
    }
    return true;
}

void ModerationEngine::ban(const std::string& player_id, const std::string& reason) {
    std::lock_guard lock(mtx_);
    bans_[player_id] = Clock::now() + std::chrono::duration_cast<Clock::duration>(
        std::chrono::duration<double>(cfg_.ban_duration_sec));
    ban_reasons_[player_id] = reason;
    std::cerr << "[mod] MANUAL BAN player=" << player_id << " reason=" << reason << "\n";
}

void ModerationEngine::unban(const std::string& player_id) {
    std::lock_guard lock(mtx_);
    bans_.erase(player_id);
    ban_reasons_.erase(player_id);
    strikes_.erase(player_id);
    windows_.erase(player_id);
}

bool ModerationEngine::is_banned(const std::string& player_id) const {
    std::lock_guard lock(mtx_);
    auto it = bans_.find(player_id);
    if (it == bans_.end()) return false;
    return Clock::now() < it->second;
}

std::vector<std::string> ModerationEngine::banned_players() const {
    std::lock_guard lock(mtx_);
    std::vector<std::string> out;
    auto now = Clock::now();
    for (auto& [id, until] : bans_) {
        if (now < until) out.push_back(id);
    }
    return out;
}

size_t ModerationEngine::strike_count(const std::string& player_id) const {
    std::lock_guard lock(mtx_);
    auto it = strikes_.find(player_id);
    return it == strikes_.end() ? 0 : it->second;
}

} // namespace ironwall

File: rust-thin-client/src/moderation.rs
//! Rate-limit + ban list (parity with C++ ModerationEngine)

use std::collections::{HashMap, VecDeque};
use std::time::{Duration, Instant};
use tracing::{info, warn};

#[derive(Debug, Clone)]
pub struct RateLimitConfig {
    pub max_events: usize,
    pub window: Duration,
    pub ban_after_violations: usize,
    pub ban_duration: Duration,
}

impl Default for RateLimitConfig {
    fn default() -> Self {
        Self {
            max_events: 30,
            window: Duration::from_secs(10),
            ban_after_violations: 3,
            ban_duration: Duration::from_secs(300),
        }
    }
}

pub struct ModerationEngine {
    cfg: RateLimitConfig,
    windows: HashMap<String, VecDeque<Instant>>,
    strikes: HashMap<String, usize>,
    bans: HashMap<String, Instant>, // until
    reasons: HashMap<String, String>,
}

impl ModerationEngine {
    pub fn new(cfg: RateLimitConfig) -> Self {
        Self {
            cfg,
            windows: HashMap::new(),
            strikes: HashMap::new(),
            bans: HashMap::new(),
            reasons: HashMap::new(),
        }
    }

    pub fn allow(&mut self, player_id: &str, event_type: &str) -> bool {
        let now = Instant::now();

        if let Some(until) = self.bans.get(player_id) {
            if now < *until {
                warn!(player=%player_id, reason=%self.reasons.get(player_id).map(|s| s.as_str()).unwrap_or(""), "DENY banned");
                return false;
            }
            self.bans.remove(player_id);
            self.reasons.remove(player_id);
        }

        let q = self.windows.entry(player_id.to_string()).or_default();
        while q.front().map(|t| now.duration_since(*t) > self.cfg.window).unwrap_or(false) {
            q.pop_front();
        }
        q.push_back(now);

        if q.len() > self.cfg.max_events {
            let s = self.strikes.entry(player_id.to_string()).or_insert(0);
            *s += 1;
            warn!(player=%player_id, event=%event_type, count=q.len(), strike=*s, "RATE LIMIT");
            if *s >= self.cfg.ban_after_violations {
                self.bans.insert(player_id.to_string(), now + self.cfg.ban_duration);
                self.reasons.insert(player_id.to_string(), "rate_limit".into());
                warn!(player=%player_id, secs=self.cfg.ban_duration.as_secs(), "BAN");
            }
            return false;
        }
        true
    }

    pub fn ban(&mut self, player_id: &str, reason: &str) {
        self.bans.insert(player_id.to_string(), Instant::now() + self.cfg.ban_duration);
        self.reasons.insert(player_id.to_string(), reason.into());
        info!(player=%player_id, reason=%reason, "MANUAL BAN");
    }

    pub fn unban(&mut self, player_id: &str) {
        self.bans.remove(player_id);
        self.reasons.remove(player_id);
        self.strikes.remove(player_id);
        self.windows.remove(player_id);
    }

    pub fn is_banned(&self, player_id: &str) -> bool {
        self.bans.get(player_id).map(|u| Instant::now() < *u).unwrap_or(false)
    }

    pub fn strike_count(&self, player_id: &str) -> usize {
        self.strikes.get(player_id).copied().unwrap_or(0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rate_limit_and_ban() {
        let cfg = RateLimitConfig {
            max_events: 3,
            window: Duration::from_secs(60),
            ban_after_violations: 2,
            ban_duration: Duration::from_secs(60),
        };
        let mut m = ModerationEngine::new(cfg);
        assert!(m.allow("a", "p"));
        assert!(m.allow("a", "p"));
        assert!(m.allow("a", "p"));
        assert!(!m.allow("a", "p"));
        assert!(!m.allow("a", "p"));
        assert!(m.is_banned("a"));
        m.unban("a");
        assert!(!m.is_banned("a"));
        assert!(m.allow("a", "p"));
    }
}
